src/lib.rs

```diff
@@ -1,171 +1,316 @@
         }
-    });
-    (timing, $key:expr, $value:expr, $rate:expr) => ({
-        if unsafe { ::std::intrinsics::unlikely($crate::within_rate(((1.0 - $rate) * ::std::u32::MAX as f64) as u32)) } {
-            let value = format!("{}", $value);
-            let rate = format!("{}", $rate);
-            $crate::send(&[$key, b":", value.as_bytes(), b"|ms", b"|@", rate.as_bytes()])
+    }
+
+    /// Query current time to use eventually with `stop_time()`
+    pub fn start_time(&self) -> StartTime {
+        StartTime( time::precise_time_ns() )
+    }
+
+    /// An efficient timer that skips querying for stop time if sample will not be collected.
+    /// Caveat : Random sampling overhead of a few ns will be included in any reported time interval.
+    pub fn stop_time(&self, key: &str, start_time: StartTime) {
+        if accept_sample(self.int_rate)  {
+            self.send_time_ms(key, start_time.elapsed_ms());
         }
-    });
-}
+    }
 
+    fn send_time_ms(&self, key: &str, interval_ms: u64) {
+        let value = &interval_ms.to_string();
+        self.send( &[key, ":", value, "|ms|@", &self.str_rate] )
+    }
 
-#[test]
-fn basics() {
-    const STATSD_SAMPLING_RATE: f64 = 0.5;
-    init("localhost:8125", "foo.bar").unwrap();
-    for _ in 0..10 {
-        statsd!(increment, b"boring", 1, STATSD_SAMPLING_RATE);
+    /// Concatenate text parts into a single buffer and send it over UDP
+    fn send(&self, strings: &[&str]) {
+        let mut str = String::with_capacity(MAX_UDP_PAYLOAD);
+        str.push_str(&self.prefix);
+        for s in strings { str.push_str(s); }
+        str.push_str(&self.str_rate);
+        self.sender.send_stats(str)
     }
+
 }
 
+/// Convert a floating point sampling rate to an integer so that a fast integer RNG can be used
+/// Float rate range is between 1.0 (send 100% of the samples) and 0.0 (_no_ samples taken)
+/// .    | float rate | int rate | percentage
+/// ---- | ---------- | -------- | ----
+/// all  | 1.0        | 0x0      | 100%
+/// none | 0.0        | 0xFFFFFFFF | 0%
+fn to_int_rate(float_rate: f64) -> u32 {
+    assert!(float_rate <= 1.0 && float_rate >= 0.0);
+    ((1.0 - float_rate) * ::std::u32::MAX as f64) as u32
+}
+
+fn accept_sample(int_rate: u32) -> bool {
+    pcg32::random() > int_rate
+}
+
+/// A convenience macro to wrap a block or an expression with a start / stop timer.
+/// Elapsed time is sent to the supplied statsd client after the computation has been performed.
+/// Expression result (if any) is transparently returned.
+#[macro_export]
+macro_rules! time {
+    ($client: expr, $key: expr, $body: block) => (
+        let start_time = $client.start_time();
+        $body
+        $client.stop_time($key, start_time);
+    );
 
-#[test]
-fn basic_behavior_of_pcg32() {
-    let mut v = Vec::new();
-    for _ in 0..100 { v.push(pcg32()) }
-    v.sort();
-    for w in v.windows(2) { assert_ne!(w[0], w[1]) }
+    ($client: expr, $key: expr, $body: expr) => (
```

Comment:
There might be some ambiguity with these two branches because the following block is also an expression:
```rust
{ 42 }
```